### ISSUE TITLE
Can't stay logged into Nextcloud for more than a minute or two of inactivity

### ISSUE THREAD

**Issue by reswild** [felixrupp/user_cas#24]
I've been testing the latest version of this app with Nextcloud 13, but unfortunately, I'm having the same problems as I had with previous versions: I can log in just fine, but I can never stay logged in for more than a minute or two of inactivity. I do get automatically logged back in again after 5-10 seconds wait as I'm using the option to force CAS login, but this is still quite inconvenient.

I've checked my logs, but I didn't find too much useful information there. There were lots of errors from the user_cas app saying "phpCAS user password has been checked", but that doesn't really sound like an error - shouldn't this be a debug message instead?

**Software:**
 - Server-OS: Debian 8
 - HTTP-Server Version: Apache 2.4.10
 - PHP-Version: PHP 5.6.33
 - Nextcloud Version: Nextcloud 13.0.4
 - user_cas-Version: 1.5.1

**Comment by felixrupp** [comment]
Hi @reswild 

the log entries of "phpCAS user password has been checked" assume, that at least one of your apps is checking the status of your oc session in a time span of a few seconds. I already have exception handling for such cases, but this does, of course, not cover all 3rd party apps yet.

If you provide a full ownCloud and Apache/PHP Log (of course anonymized), I can try to find out which app/process is triggering the session auth checking and can then blacklist this call.

Regards
Felix

**Comment by reswild** [comment]
The "phpCAS user password has been checked" message only happens late in the process, after I have not been able to load a page, have gotten the "Problem loading page. Reloading in ... seconds" message in my browser, and is then apparently being logged in twice by CAS. For some reason, everything in the user_cas app seems to be executed twice.

The only messages in my log at the time I'm not able to load a page is one or two lines that seems to originate from remote.php. First I get a line with "Login failed", and then one with "Sabre\DAV\Exception\NotAuthenticated: Cannot authenticate over ajax calls".

I have switched this site back to using the user_saml app, but if I can find the time, I'll try doing a fresh install of Nextcloud on one of my other servers so I can test this some more, and then report back to you.

**Comment by felixrupp** [comment]
@reswild That’s exactly the behaviour which occurs, when other apps trigger controller actions which need a user to be logged in and user_cas does not know about how the app works. Most of these issues occur with ajax-based calls. Can you post the entries of your browser developer console, where you can see which call first gets the „failed“ answer back? I would need to see the whole script-path with parameters to exclude this type of calls from user_cas handling.

The remote.php handling should already be excluded from user_cas handling, but that has been tested mostly with the ownCloud 10 version. I don’t know how different the Nextcloud 13 remote.php handling is from the oC one.

Regards
Felix

**Comment by reswild** [comment]
Well, I tried installing Nextcloud 13 on one of my other servers, this one running Nginx and PHP 7.0. Two things happened when I activated the user_cas app:
1. user_cas started filling up my log with multiple debug entries every second, all of them saying "phpCAS Nextcloud 13.0.4. detected."
2. If I tried to log in (whether or not I was trying to use CAS), I got a WSOD with the message "Error: Internal script failure"
I suppose this could be an issue with my nginx configuration, but all other apps seems to work fine.

I did also try to activate CAS again on the original site, and check my console. The only errors I'm seeing here are '401 Unauthorized' errors on PROPFIND requests to remote.php/webdav/[folder_name].

With my regular set-up, I was also getting 
`Refused to connect to 'https://example.com/cas/login?service=https%3A%2F%2Fcloud.example.com%2Findex.php%2Fapps%2Fuser_cas%2Flogin' because it violates the following Content Security Policy directive: "connect-src 'self'".`
However, if I move the Nextcloud installation to the same subdomain as my login server, this error goes away but I'm still being logged out same as before.

**Comment by felixrupp** [comment]
Hi @reswild 

1. This is expected behaviour at the moment, because your log level is probably set to 'All' (0). user_cas is a quite chatty app, that means, it makes excessive use of log-writes, to help debug CAS problems (because most CAS servers are black boxes). To avoid that, please raise your log level in Nextcloud config.php to at least 2 or 3.

2. Can you post the according "Internal Server Error" from the servers error log file? That would possibly show where the error occured.

The „Content Security Policy“ errors are thrown by javascripts, which try to access the CAS server, just because user_cas is not filtering them at the time. This is an error which can be solved, if you post a detailed http-access-log from your browsers developer console. I can add them to the blacklist then, and the errors will be gone.

Regards,
Felix

**Comment by reswild** [comment]
Surely it's not expected behaviour that the user_cas app should flood my log with debug messages 3 to 5 times a second, all notifying me about "phpCAS Nextcloud 13.0.4. detected"? I usually set the log level to 'errors only' on my production sites, but I need to be able to set this to debug when testing stuff on my dev sites.

I did some debugging of my WSOD error, and I found that the cause for this was a "bad CAS server port" error from phpCAS. Specifying the port number in the user_cas configuration solved this. (This field had a placeholder saying "443" so I had assumed that meant that 443 was already set as the default portnumber, and I therefore didn't need to fill in this field.)

So now I can log in using CAS on this server also, but I'm having the same problem of getting logged out again after a short time.

According to my console, the script that is called immediately before I get my „Content Security Policy“ error is index.php/apps/files/ajax/getstoragestats.php.

So where in your code is this blacklist you are mentioning? I'm not sure if I understand how this is working.

**Comment by reswild** [comment]
OK, so I finally set aside some time to look into this properly, and the reason I keep getting logged out appears to be that the user_cas app is writing both a password and a token to the authtoken table. This means that Nextcloud tries to verify that I'm logged in by checking the password instead of the token, with the result that I can't stay logged in for more than 5 minutes. Replacing the password with a NULL value in the authtoken table makes me stay logged in.

I found two lines in the appinfo/app.php file that excludes various pages from running the code here - I guess this is the blacklist you were referring to. To be honest, I don't think this is the best approach. Wouldn't it be better to do this the other way abound and instead list the pages you do want to run the code? Probably the redirect code only needs to be run on the login page.

I also found that the reason the log got flooded with debug messages was that I was monitoring the log page while testing stuff, and this caused the user_cas code to run every time the log was polled. I added `|| strpos($requestUri, 'logreader')` to your blacklist, and this got the logging back to normal again.
